File: 11-20/0015 | 3Sum/solution.1.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> stock;
        
        if(nums.size()<3)
            return {};
        
        for(int i=0;i<nums.size()-2;i++){
            if(i==0 || (i>0 && nums[i] != nums[i-1])){
                int target = -nums[i];
                int low = i+1;
                int high = nums.size()-1;
                while(low<high){
                    if(nums[low]+nums[high] == target){
                        stock.push_back({nums[i],nums[low],nums[high]});
                        while(low<high && nums[low]==nums[low+1])
                            low++;
                        while(low<high && nums[high]==nums[high-1])
                            high--;
                        low++;
                    }
                    else if(nums[low]+nums[high] < target)
                        low++;
                    else
                        high--;
                }
            }
        }
        return stock;
        
    }

};

File: 11-20/0018 | 4Sum/solution.1.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    let stock = []
    nums.sort((a,b) => a-b)
    for(let i=0;i<nums.length-3;i++){
        if(i>0 && nums[i] == nums[i-1]){
                continue;
            }
        for(let j=i+1;j<nums.length-2;j++){
            if(j>i+1 && nums[j] == nums[j-1]){
                continue;
            }
            
                let start = j+1
                let end = nums.length-1
                let sumA = nums[i] + nums[j]
                
                while(start < end){
                    let sumB = nums[start] + nums[end] + sumA
                    
                    if(sumB == target){
                        let t = [nums[i],nums[j],nums[start],nums[end]]
                        stock.push(t)
                        
                        start++
                        end--
                        
                        while(start < end && (nums[start] == nums[start-1])){
                            start++
                        }
                        
                        while(start < end && (nums[end] == nums[end+1])){
                            end--
                        }
                    }
                    
                    else if (sumB < target){
                        start++
                    }
                    
                    else {
                        end--
                    }
                    
                }
            } 
    }
    return stock
};
